### ISSUE TITLE
genetic algos: how to export more than just the goals of a optimisation?

### ISSUE THREAD

**Issue by samthiriot** [openmole/openmole#483]
Dear beloved OpenMOLE community,

when we run a model M, it typically produces many outputs; only a few are used more or less directly as goals for the optimization algorithm. 

I understood how to use the `SavePopulationHook` to dump the best individuals of the current generation.
This only dumps the point of view of the optimization algorithm, that is the inputs and outputs manipulated by the optimization algo. 

Is it possible to export as well other outputs of the model, so we can check what happens on these outputs, even if they are not directly constrained by the algo?

I tried to add a CSVHook to the execution of the model, but this looses track of the current generation. 

I provide below a simple example as a basis of discussion. In this example, I'm trying to get a CSV file containing the generation, `x`, `f1`, `f2` and `z`. 

```
// explored space
val A = 1000.0 // up to 10^5

// the test function 
val testFunctionSchafferN1 =
    ScalaTask("""
        val f1 = Math.pow(x, 2);
        val f2 = Math.pow(x - 2, 2);
        val z = f1 + f2
        """) set (
    (inputs, outputs) += x,
    outputs += f1,
    outputs += f2,
    outputs += z
    )
    
// the optimisation algorithm under test 
val evolutionSchafferN1 =
  NSGA2Evolution(
    genome = Seq(
      x in (-A, A)
    ),
    objective = Seq(f1, f2),
    evaluation = testFunctionSchafferN1, 
    parallelism = 4,
    termination = 100
  ) hook DisplayHook()
  
// compute evolution on the test Function
evolutionSchafferN1 on envMultiThread hook SavePopulationHook(evolutionSchafferN1, workDirectory/"exported.csv")
```

Is there any way to do it with the existing components? If I have to write it myself, should it take the direction of a hook based on `SavePopulationHook`?

Thanks!

Sam

**Comment by romainreuillon** [comment]
Dear Sam,

this is not possible at the moment (I was thinking of reevaluating if it is the right choice in a near future). 

The logic is that GA are often heavy on memory in openmole, therefore we keep the minimum of info for each run and once acceptable solution have been found, it is possible to them in depth using a CSVSampling.

For instance if you run a GA with islands and stochasticity the number of evaluation kept in memory (a part of it offloaded to disc to allow better scalability) : individuals x islands x replications.

The computing time of the analysis a posteriori is negligible compared to the computing time spent to compute the results of the GA.

Romain

**Comment by JusteRaimbault** [comment]
Kind of a coincidence, we were discussing exactly that last Thursday during the summer school - indeed we needed to have additional covariates which were not objectives as an output of PSE.
One solution to make it transparent to the user would be to keep the seeds and integrate the recomputation for the final population only as an option in the evolution task?

**Comment by samthiriot** [comment]
Thanks @romainreuillon for the detailed answer,
Thanks @JusteRaimbault for the additional elements,

I think it is sad not to be able to keep more results of the evolution. I tried to list the arguments matching general use cases:
* replaying in not yet possible in the case of stochastic optimization. As pointed out by @JusteRaimbault the seed is not yet exported. So in the case of a "strongly" stochastic model, it is difficult to reproduce the simulations of interest. Exporting the seed in the hook seems difficult as there are several seeds per individual (as many as samples indeed). 
* in the case the simulations are slow, loosing the results of simulations is a loss on several aspects: time, computational effort, energy consumption. Imagine a simulation which requires 1h to run; every time we want to analyze one point of one PAreto front during the evolution, was have to manually set up the corresponding parameters, run it, wait for 1h. It also requires an additional manual step which is error prone, as any manual step. 
* in several cases we can not evaluate the quality of convergence without looking at more detailed results. We work for instance on a case where the simulation is producing a time serie (in this case, a load curve of energy consumption), and optimization tries to reduce the distance between the simulate time serie and a reference time serie. The goal of the optimization is aggregated (for instance, distance between the two load curves), so just looking at the results of the goal is not informative. What is really required to understand the quality of results is to watch the load curve produced by the algorithm. My point is: exporting more data is not required only when we are satisfied by the results, but during the entire iterative process of calibration.

Pleasure to discuss it more!

Sam

**Comment by samthiriot** [comment]
Hello all,

I possibly found a partial workaround by using a CSV hook on the model
```

//model inputs
val x = Val[Double]
val y = Val[Double]

//model outputs and goals
val f1 = Val[Double] 
val f2 = Val[Double]
// other model outputs
val f3 = Val[Double] 

// the test function 
val testFunctionBinhKorn =
    ScalaTask("""
        val g1 = Math.pow(x - 5, 2) + Math.pow(y, 2)
        val g2 = Math.pow(x - 8, 2) + Math.pow(y + 2, 3)
        val constrained = g1 > 25 || g2 < 7.7
        
        val f1 = if (constrained) Double.NaN else 4*Math.pow(x,2) + 4*Math.pow(y,2)
        val f2 = if (constrained) Double.NaN else Math.pow(x-5,2) + Math.pow(y-5,2)
        
        val f3 = 42 + f1 + f2
        """) set (
    inputs += x,
    inputs += y,
    outputs += f1,
    outputs += f2,
    outputs += f3
    )
    
// the optimisation algorithm under test 
val evolutionBinhKorn =
  NSGA2Evolution(
    genome = Seq(
      x in (0.0, 5.0),
      y in (0.0, 3.0)
    ),
    objective = Seq(f1, f2),
    evaluation = testFunctionBinhKorn hook CSVHook(workDirectory / "detailed.csv"),
    parallelism = 5,
    termination = 1000
  )
```

In this example the file "detailed.csv" contains all the outputs of the simulation, including an output which is not a goal.
To add the inputs as well, we need to forward them as outputs:
```
// the test function 
val testFunctionBinhKorn =
    ScalaTask("""
        val g1 = Math.pow(x - 5, 2) + Math.pow(y, 2)
        val g2 = Math.pow(x - 8, 2) + Math.pow(y + 2, 3)
        val constrained = g1 > 25 || g2 < 7.7
        
        val f1 = if (constrained) Double.NaN else 4*Math.pow(x,2) + 4*Math.pow(y,2)
        val f2 = if (constrained) Double.NaN else Math.pow(x-5,2) + Math.pow(y-5,2)
        
        val f3 = 42 + f1 + f2
        """) set (
    (inputs, outputs) += x,
    (inputs, outputs) += y,
    outputs += f1,
    outputs += f2,
    outputs += f3
    )
```
Then we have inputs and outputs for every run. I bet the seed can be managed as any other input so it can be exported. 

What I guess we miss there is:
* the iteration of the evolution,
* we don't have the aggregate quality of the simulation any more

Maybe if it was possible to export an ID of individual for every run, and from the Evolution hook, we would be able to later join the two result files with a simple postprocessing... 

what do you think? 

Sam

**Comment by romainreuillon** [comment]
Dear Sam,

I have to work on some stuff before diving into this aspect but it is on my todo list. I think we could discuss about what you missed then in order to be sure I understand the need properly.

I can just point 2 reasoning that I think are flawed:
 - if the fitness function evaluation last for 1h, the computation of the Pareto front probably last for several weeks. Running a direct sampling to post process the result for 1h is still negligible compared to the huge computation load of the evolution process (let say at top 1/1000 of the evolution computation load). 
 - for a stochastic model if the result is very uncommon and doesn't reflect an effect of the input parameters, it is likely to be evicted from an evolution method which is robust to stochastic (evolution method are about finding inputs that lead to some kind of optimal solution). If this is what is sought, a way to do that is to maximize some kind of variance estimation and then post-process to get some good representation of the distribution of the random variate for a given input.

All that said, I agree that it might be convenient to add the possibility to save some outputs other than the objectives in the evolution process. For me the motivation is mostly to improve the ergonomic design of these methods (which is an important aspect of OpenMOLE).

About the probability of success, I am not able to answer yet. I have to dive into the code and experiment to check what is possible (especially for the case of stochastic fitness functions).

cheers,
Romain

**Comment by romainreuillon** [comment]
The possibility to include additional outputs in the result file has been implemented in : 1c9125c09150fe6e6ed3be6cc8ada5abc5695951

This functionality is experimental for now and may be subject to changes. 

To use it add `includeOutputs = true` to the hook:
```
) hook (..., includeOutputs = true)
```
